### PR TITLE
Add last_updated_at and total_updates to datasets list view

### PR DESCRIPTION
**Pull request by blag**
This PR fleshes out the dataset summary view, displaying the last update datetime of each dataset.

The endpoint can also sort by `last_dataset_update` (latest first or oldest first) or by the dataset URI (ascending or descending). ~By default, datasets without any updates are displayed first, and then datasets with the latest updates are displayed (eg: nulls first, descending by update time). I pushed the nulls to the top because I think it might be more useful for people who are investigating why a DAG is not being triggered by a dataset, so it might be more helpful to have datasets that have never been updated as easily accessible as possible. However, if this is a contentious choice, I'm happy to revert that change.~ Reverted.

Other authors, including @jedcunningham, @ephraimbuddy, and @bbovenzi did most of the groundwork on this, I just added some final touches and fixed a few bugs.

## Before ##
dataset_list_before

## After ##
dataset_list_after

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by ashb** [comment]
Could you include before/after screenshots in the PR desc please?

**Comment by blag** [comment]
Updated description with screenshots and giving credit where credit is due.

File: airflow/www/views.py
Comment (jedcunningham) [inline airflow/www/views.py]:
```suggestion
```

File: airflow/www/views.py
Comment (jedcunningham) [inline airflow/www/views.py:3589]: How does one actually hit this block via the UI?

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
        clear_db_datasets()
```

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
```

Probably don't need this.

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
```

Same here.

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
```

And here.

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
                uri=f"s3://bucket/key/{i}",
```

Maybe just stick with this for consistency sake?

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py:124]:
```suggestion
        session.add_all([
            DatasetEvent(
                dataset_id=datasets[2].id,
                timestamp=pendulum.today('UTC').add(days=i),
            ) for i in [-3, -2, -1]
        ])
```

nit: might need some formatting still

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
            session.add_all([
                DatasetEvent(
                    dataset_id=ds1_id,
                    timestamp=pendulum.DateTime(2022, 8, 1, tzinfo=UTC),
                ) for i in range(3)
            ])
```

nit: similar to above.

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
```

Not sure we need these?

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
```

Same here.

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
```

These too?

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
            DatasetModel(uri=f"s3://bucket/key/{i}") for i in range(1, 10)
```

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
            DatasetModel(uri=f"s3://bucket/key/{i}") for i in range(1, 60)
```

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [inline tests/www/views/test_views_dataset.py]:
```suggestion
            DatasetModel(uri=f"s3://bucket/key/{i}") for i in range(1, 60)
```

File: airflow/www/views.py
Comment (blag) [reply airflow/www/views.py:3589]: The [default order](https://github.com/apache/airflow/pull/26358/files#diff-5ed4b421389b931608cae97949c8e129aa9d4076e52d71a9e7f1b42ff49f77a5R75) is `-last_dataset_update`, but we do not yet have a way for users to explicitly sort this way in the UI.

I can attempt to add this in, or we can add it in later.

File: tests/www/views/test_views_dataset.py
Comment (blag) [reply tests/www/views/test_views_dataset.py]: Disagree. The [select and join](https://github.com/apache/airflow/pull/26358/files#diff-917b70d2661fae8322c538a4f60b9d4ba755556036fe0c527eb0411a6ac60f2eR3628-R3650) are not simple, and they rely on aggregating timestamps on related objects. I explicitly included these timestamps here to help keep any errors made in the query loud, instead of potentially silent.

But I'm happy to remove them if you insist.

File: airflow/www/views.py
Comment (jedcunningham) [reply airflow/www/views.py:3589]: Part of why I asked is everything following is pretty duplicative for asc/desc. An easy "fix" was to just not have it :)

File: tests/www/views/test_views_dataset.py
Comment (blag) [reply tests/www/views/test_views_dataset.py:124]: These three dataset events point to two different datasets (`datasets[2]` and `datasets[1]`), so a simple loop is not equivalent.

And I figured a loop for the two dataset events for `datasets[1]` was probably overkill.

File: tests/www/views/test_views_dataset.py
Comment (jedcunningham) [reply tests/www/views/test_views_dataset.py]: I mean, this test specifically should never even get that far? Plus in general these timestamps on DatasetModel aren't actually used in the query at all?

File: airflow/www/views.py
Comment (blag) [reply airflow/www/views.py:3589]: It's absolutely disappointingly messy, that's for sure. 😕

I do think that supporting both asc/desc is useful for answering different questions, but I'm not 100% sure if the nulls first/last for the respective ordering is the most optimal, especially for deployments with a great many datasets. Totally willing to nix and fix the pairings if we don't like that.

File: tests/www/views/test_views_dataset.py
Comment (blag) [reply tests/www/views/test_views_dataset.py]: > I mean, this test specifically should never even get that far?

A fair point. :+1:

> Plus in general these timestamps on DatasetModel aren't actually used in the query at all?

While this is true, I think well written fixtures are also useful for trying to catch things that future contributors might mess up. And mistyping something to use `DatasetModel.created_at` instead of `DatasetDagRunQueue.created_at`, for example, is one of them (`updated_at` less so, but I figure it can't really hurt anything other than minute readability of the test code, and something else might have an `updated_at` attribute/column added in the future).

But this is also speculative - I'll remove it, here and elsewhere where it isn't needed.

File: airflow/www/views.py
Comment (blag) [reply airflow/www/views.py:3589]: If we do the categorizing and sorting in the UI a la [this excellent suggestion](https://github.com/apache/airflow/issues/26381#issuecomment-1248227417), we can avoid doing the messy ordering in the SQLA query altogether. 🤔

Edit: Hmm, but then it becomes practically impossible for the `limit` parameter to be useful at all. 😕

Comment (bbovenzi) [reply airflow/www/views.py:3589]: We may want to sort in a couple ways right? By name and by last updated at? I think having the UI just default to desc was so that the latest changes are what a user sees first.

Comment (bbovenzi) [reply airflow/www/views.py:3589]: Oh also, a blank last updated should probably display after datasets with updates instead of before.

**Comment by bbovenzi** [comment]
Actually, I'm don't think otal updates count is even useful. I'd say lets just remove it entirely and have this PR just be about adding the last updated at info. Or put it into draft until I talk to some users and come up with a better solution

**Comment by blag** [comment]
@bbovenzi I fixed the issues with the aggregate counts (on PostgreSQL) and expanded the tests a little bit to include more DAG <-> dataset relationships to exercise those SQL paths. Still need to fix things on...the rest of the supported database backends.

**Comment by bbovenzi** [comment]
I changed last updated to its own column that the user can sort by.

Screen Shot 2022-09-27 at 9 37 00 AM

File: airflow/www/views.py
Comment (dstandish) [inline airflow/www/views.py]: do these two left joins do anything?

File: airflow/www/views.py
Comment (dstandish) [inline airflow/www/views.py]: given that this query will be more expensive than just querying the datasets table, maybe it makes sense to call this like .... `get_datasets_summary` or `get_dataset_summary_stats` or something.... because if all you want to do is list the datasets, maybe you wouldn't need or expect the other stuff

File: airflow/www/views.py
Comment (dstandish) [inline airflow/www/views.py:3589]: i don't know if you are supporting pagination here but... or if it matters, but, if you are ordering by dataset event timestamp, you might get unexpected results cus the order might change, with new events being added.  probably the timeframes are short enough that it wouldn't really matter.

**Review by bbovenzi — APPROVED** [review]
Testing this all out locally. Looks good to me. But we should have another review since I wrote much of the UI code and I'm not too familiar with all the special session query changes.

**Review by dstandish — DISMISSED** [review]

**Review by dstandish — COMMENTED** [review]
sorry meant to just comment 🙃

File: airflow/www/views.py
Comment (blag) [reply airflow/www/views.py]: Not anymore; removed, thanks!

File: airflow/www/views.py
Comment (bbovenzi) [reply airflow/www/views.py]: +1 Let's differentiate the name from the REST API endpoint.

File: airflow/www/views.py
Comment (bbovenzi) [reply airflow/www/views.py:3589]: This does have pagination but we don't do any auto-refreshing, at least for now.

File: airflow/www/views.py
Comment (blag) [reply airflow/www/views.py]: Renamed to `/object/datasets_summary`.

Comment (bbovenzi) [reply airflow/www/views.py]: should the function name also be changed to match the url? stay consistent?

Comment (blag) [reply airflow/www/views.py]: Fixed; changed to `datasets_summary`.

File: airflow/www/templates/airflow/datasets.html
Comment (bbovenzi) [inline airflow/www/templates/airflow/datasets.html]:
```suggestion
  <meta name="datasets_api" content="{{ url_for('Airflow.datasets_summary') }}">
```

Need to update this to match the function name in `views.py`

**Comment by ldacey** [comment]
I haven't had a chance to move to 2.4.0 yet. Is there a way to filter datasets by tags, potentially? I normally tag different dags with the project/client name. It would be cool to filter for a single "project" (consisting of multiple DAGs) and see all of the datasets and how they interconnect.

**Comment by blag** [comment]
> I haven't had a chance to move to 2.4.0 yet. Is there a way to filter datasets by tags, potentially?

No, we haven't implemented tagging datasets yet. That's a good idea though.

> see all of the datasets and how they interconnect

This page does have a good dataset/dag dependency visualization, so definitely check that out once you upgrade to 2.4!

**Review by bbovenzi — APPROVED** [review]

**Comment by bbovenzi** [comment]
> I haven't had a chance to move to 2.4.0 yet. Is there a way to filter datasets by tags, potentially? I normally tag different dags with the project/client name. 

After this PR merges, I plan to add more searching and filtering. Adding tags could be a good idea.

> It would be cool to filter for a single "project" (consisting of multiple DAGs) and see all of the datasets and how they interconnect.
Yes, that is also on my to-do list. Right now the page can get too busy very quickly.

**Comment by ldacey** [comment]
Nice, that would be great. A lot of our clients are completely distinct - if I am showing a demo or screenshot for "Client A", I do not want to show "Client B" at all, and I might want to see how all datasets interconnect for "Client A". There might be other approaches, but I have used tags to filter my DAGs UI view historically.

**Pull request merged**
